shell/platform/darwin/ios/framework/Source/FlutterMetalLayer.mm

```diff
@@ -0,0 +1,402 @@
+      if (_framebufferOnly) {
+        textureDescriptor.usage = MTLTextureUsageRenderTarget;
+      } else {
+        textureDescriptor.usage =
+            MTLTextureUsageRenderTarget | MTLTextureUsageShaderRead | MTLTextureUsageShaderWrite;
+      }
+      id<MTLTexture> texture = [self.device newTextureWithDescriptor:textureDescriptor
+                                                           iosurface:(__bridge IOSurfaceRef)surface
+                                                               plane:0];
+      FlutterTexture* flutterTexture = [[FlutterTexture alloc] initWithTexture:texture
+                                                                       surface:surface];
+      return flutterTexture;
+    } else {
+      // Make sure raster thread doesn't have too many drawables in flight.
+      if (_availableTextures.count == 0) {
+        CFTimeInterval start = CACurrentMediaTime();
+        while (_availableTextures.count == 0 && CACurrentMediaTime() - start < 1.0) {
+          usleep(100);
+        }
+        CFTimeInterval elapsed = CACurrentMediaTime() - start;
+        if (_availableTextures.count == 0) {
+          NSLog(@"Waited %f seconds for a drawable, giving up.", elapsed);
+          return nil;
+        } else {
+          NSLog(@"Had to wait %f seconds for a drawable", elapsed);
+        }
+      }
+      // Return first drawable that is not in use or the one that was presented
+      // the longest time ago.
+      FlutterTexture* res = nil;
+      for (FlutterTexture* texture in _availableTextures) {
+        if (!texture.surface.isInUse) {
+          res = texture;
+          break;
+        }
+        if (res == nil || texture.presentedTime < res.presentedTime) {
+          res = texture;
+        }
+      }
+      [_availableTextures removeObject:res];
+      return res;
+    }
+  }
+}
+
+- (id<CAMetalDrawable>)nextDrawable {
+  FlutterTexture* texture = [self nextTexture];
+  if (texture == nil) {
+    return nil;
+  }
+  FlutterDrawable* drawable = [[FlutterDrawable alloc] initWithTexture:texture
+                                                                 layer:self
+                                                            drawableId:_nextDrawableId++];
+  return drawable;
+}
+
+- (void)presentOnMainThread:(FlutterTexture*)texture {
+  // This is needed otherwise frame gets skipped on touch begin / end. Go figure.
+  // Might also be placebo
+  [self setNeedsDisplay];
+
+  [CATransaction begin];
+  [CATransaction setDisableActions:YES];
+  self.contents = texture.surface;
+  texture.presentedTime = CACurrentMediaTime();
+  [CATransaction commit];
+  _displayLink.paused = NO;
+  _displayLinkPauseCountdown = 0;
+  if (!_didSetContentsDuringThisDisplayLinkPeriod) {
+    _didSetContentsDuringThisDisplayLinkPeriod = YES;
+  } else if (!_displayLinkForcedMaxRate) {
+    _displayLinkForcedMaxRate = YES;
+    [self setMaxRefreshRate:[DisplayLinkManager displayRefreshRate] forceMax:YES];
```

Comment:
why are we changing the refresh rate here?

Reply:
There is a weird behavior on iPhone, where for several seconds after unlock, and only with variable refresh rate enabled, the main thread display link seems to fire on completely different interval than actual system refresh rate. This does not happen when non-variable refresh rate (hence forcing the refresh rate if the layer detects multiple buffers submitted during single display link interval).

It is documented after top of the class (perhaps not the best place):
```cpp
  // Used to track whether the content was set during this display link.
  // When unlocking phone the layer (main thread) display link and raster thread
  // display link get out of sync for several seconds. Even worse, layer display
  // link does not seem to reflect actual vsync. Forcing the layer link
  // to max rate (instead range) temporarily seems to fix the issue.
  BOOL _didSetContentsDuringThisDisplayLinkPeriod;
```

This behavior can be observed through animation hitches instruments tool for example - after modifying `vsync_waiter_ios` to fire on main thread, you will see  that UI and raster threads are bing woken up at completely different period than system vsync.

It's strage that this happens and the only workaround I found was to disable variable refresh rate for a bit once this starts happening.